src/library/blas3/clsparse-spm-spm.cpp

```diff
@@ -0,0 +1,1009 @@
+/* ************************************************************************
+* The MIT License (MIT)
+* Copyright 2014-2015 weifengliu
+*  Permission is hereby granted, free of charge, to any person obtaining a copy
+*  of this software and associated documentation files (the "Software"), to deal
+*  in the Software without restriction, including without limitation the rights
+*  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+*  copies of the Software, and to permit persons to whom the Software is
+*  furnished to do so, subject to the following conditions:
+
+*  The above copyright notice and this permission notice shall be included in
+*  all copies or substantial portions of the Software.
+
+*  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+*  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+*  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+*  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+*  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+*  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+*  THE SOFTWARE.
+* ************************************************************************ */
+
+/* ************************************************************************
+*  < A CUDA/OpenCL General Sparse Matrix-Matrix Multiplication Program >
+*
+*  < See papers:
+*  1. Weifeng Liu and Brian Vinter, "A Framework for General Sparse
+*      Matrix-Matrix Multiplication on GPUs and Heterogeneous
+*      Processors," Journal of Parallel and Distributed Computing, 2015.
+*  2. Weifeng Liu and Brian Vinter, "An Efficient GPU General Sparse
+*      Matrix-Matrix Multiplication for Irregular Data," Parallel and
+*      Distributed Processing Symposium, 2014 IEEE 28th International
+*      (IPDPS '14), pp.370-381, 19-23 May 2014.
+*  for details. >
+* ************************************************************************ */
+
+
+#include "include/clSPARSE-private.hpp"
+#include "internal/clsparse-control.hpp"
+#include "internal/kernel-cache.hpp"
+#include "internal/kernel-wrap.hpp"
+#include "internal/clsparse-internal.hpp"
+
+#include <cmath>
+
+#define GROUPSIZE_256 256
+#define TUPLE_QUEUE 6
+#define NUM_SEGMENTS 128
+#define WARPSIZE_NV_2HEAP 64
+#define value_type float
+#define index_type int 
+#define MERGEPATH_LOCAL     0
+#define MERGEPATH_LOCAL_L2  1
+#define MERGEPATH_GLOBAL    2
+#define MERGELIST_INITSIZE 256
+#define BHSPARSE_SUCCESS 0
+ 
+#define CHECKMAL(a, b) \
+    if(a==NULL) \
+    {     \
+        std::cerr << "Malloc Error:" << b << std::endl; \
+        exit(1); \
```

Comment:
Remove any and all occurrences of exit() from the library.  Libraries should never terminate a process; we need to return an error code.